File: services/config/src/flags/evaluator.cpp
#include "apdl/flags/evaluator.h"
#include "spdlog/spdlog.h"
#include "rapidjson/writer.h"
#include "rapidjson/stringbuffer.h"

#include <functional>
#include <cmath>
#include <algorithm>
#include <regex>

namespace apdl {

// FNV-1a hash for consistent bucketing. This is deterministic and produces
// a uniform distribution, making it suitable for percentage-based rollouts
// and experiment variant assignment.
uint32_t FlagEvaluator::hashBucket(const std::string& key, const std::string& user_id) {
    std::string input = key + ":" + user_id;

    // FNV-1a 32-bit
    uint32_t hash = 2166136261u;
    for (char c : input) {
        hash ^= static_cast<uint32_t>(static_cast<unsigned char>(c));
        hash *= 16777619u;
    }
    return hash;
}

bool FlagEvaluator::isInRollout(const std::string& flag_key, const std::string& user_id,
                                 double percentage) {
    if (percentage >= 100.0) return true;
    if (percentage <= 0.0) return false;

    uint32_t hash = hashBucket(flag_key, user_id);
    // Map hash to 0-100 range
    double bucket = (static_cast<double>(hash) / static_cast<double>(UINT32_MAX)) * 100.0;
    return bucket < percentage;
}

std::string FlagEvaluator::selectVariant(const std::string& flag_key, const std::string& user_id,
                                          const std::string& variants_json) {
    rapidjson::Document doc;
    doc.Parse(variants_json.c_str(), variants_json.size());

    if (doc.HasParseError() || !doc.IsArray() || doc.GetArray().Size() == 0) {
        return "";
    }

    const auto& variants = doc.GetArray();

    // Calculate total weight
    double total_weight = 0.0;
    for (rapidjson::SizeType i = 0; i < variants.Size(); ++i) {
        if (variants[i].IsObject() && variants[i].HasMember("weight")) {
            total_weight += variants[i]["weight"].GetDouble();
        }
    }

    if (total_weight <= 0.0) {
        // Equal weights if none specified
        total_weight = static_cast<double>(variants.Size());
    }

    uint32_t hash = hashBucket(flag_key + ":variant", user_id);
    double bucket = (static_cast<double>(hash) / static_cast<double>(UINT32_MAX)) * total_weight;

    double cumulative = 0.0;
    for (rapidjson::SizeType i = 0; i < variants.Size(); ++i) {
        if (!variants[i].IsObject()) continue;

        double weight = 1.0; // default equal weight
        if (variants[i].HasMember("weight") && variants[i]["weight"].IsNumber()) {
            weight = variants[i]["weight"].GetDouble();
        }

        cumulative += weight;
        if (bucket < cumulative) {
            // Return this variant's value
            if (variants[i].HasMember("value")) {
                if (variants[i]["value"].IsString()) {
                    return variants[i]["value"].GetString();
                }
                // Serialize non-string values
                rapidjson::StringBuffer sb;
                rapidjson::Writer<rapidjson::StringBuffer> w(sb);
                variants[i]["value"].Accept(w);
                return std::string(sb.GetString(), sb.GetSize());
            }
            if (variants[i].HasMember("key") && variants[i]["key"].IsString()) {
                return variants[i]["key"].GetString();
            }
            return std::to_string(i);
        }
    }

    // Fallback to last variant (should not happen with proper weights)
    return "";
}

bool FlagEvaluator::matchesCondition(const rapidjson::Value& condition, const EvalContext& ctx) {
    if (!condition.IsObject()) return false;

    // A condition has: attribute, operator, value
    if (!condition.HasMember("attribute") || !condition["attribute"].IsString()) return false;
    if (!condition.HasMember("operator") || !condition["operator"].IsString()) return false;

    std::string attribute = condition["attribute"].GetString();
    std::string op = condition["operator"].GetString();

    // Look up the attribute value from the context
    std::string actual_value;
    if (attribute == "user_id" || attribute == "userId") {
        actual_value = ctx.user_id;
    } else if (attribute == "anonymous_id" || attribute == "anonymousId") {
        actual_value = ctx.anonymous_id;
    } else {
        auto it = ctx.attributes.find(attribute);
        if (it != ctx.attributes.end()) {
            actual_value = it->second;
        } else {
            // Attribute not found -- most operators should not match
            if (op == "not_exists" || op == "is_not_set") return true;
            return false;
        }
    }

    if (!condition.HasMember("value")) {
        // Some operators don't need a value (exists, not_exists)
        if (op == "exists" || op == "is_set") return !actual_value.empty();
        if (op == "not_exists" || op == "is_not_set") return actual_value.empty();
        return false;
    }

    // Get the expected value(s) from the condition
    const auto& expected = condition["value"];

    if (op == "equals" || op == "eq" || op == "is") {
        if (expected.IsString()) return actual_value == expected.GetString();
        if (expected.IsBool()) return actual_value == (expected.GetBool() ? "true" : "false");
        if (expected.IsInt()) return actual_value == std::to_string(expected.GetInt());
        if (expected.IsDouble()) return actual_value == std::to_string(expected.GetDouble());
        return false;
    }

    if (op == "not_equals" || op == "neq" || op == "is_not") {
        if (expected.IsString()) return actual_value != expected.GetString();
        if (expected.IsBool()) return actual_value != (expected.GetBool() ? "true" : "false");
        return true;
    }

    if (op == "contains") {
        if (expected.IsString()) {
            return actual_value.find(expected.GetString()) != std::string::npos;
        }
        return false;
    }

    if (op == "not_contains") {
        if (expected.IsString()) {
            return actual_value.find(expected.GetString()) == std::string::npos;
        }
        return true;
    }

    if (op == "starts_with") {
        if (expected.IsString()) {
            std::string prefix = expected.GetString();
            return actual_value.size() >= prefix.size() &&
                   actual_value.substr(0, prefix.size()) == prefix;
        }
        return false;
    }

    if (op == "ends_with") {
        if (expected.IsString()) {
            std::string suffix = expected.GetString();
            return actual_value.size() >= suffix.size() &&
                   actual_value.substr(actual_value.size() - suffix.size()) == suffix;
        }
        return false;
    }

    if (op == "in") {
        if (expected.IsArray()) {
            for (rapidjson::SizeType i = 0; i < expected.GetArray().Size(); ++i) {
                if (expected[i].IsString() && actual_value == expected[i].GetString()) {
                    return true;
                }
            }
        }
        return false;
    }

    if (op == "not_in") {
        if (expected.IsArray()) {
            for (rapidjson::SizeType i = 0; i < expected.GetArray().Size(); ++i) {
                if (expected[i].IsString() && actual_value == expected[i].GetString()) {
                    return false;
                }
            }
        }
        return true;
    }

    // Numeric comparisons
    if (op == "gt" || op == "gte" || op == "lt" || op == "lte") {
        double actual_num, expected_num;
        try {
            actual_num = std::stod(actual_value);
        } catch (...) {
            return false;
        }

        if (expected.IsNumber()) {
            expected_num = expected.GetDouble();
        } else if (expected.IsString()) {
            try {
                expected_num = std::stod(expected.GetString());
            } catch (...) {
                return false;
            }
        } else {
            return false;
        }

        if (op == "gt") return actual_num > expected_num;
        if (op == "gte") return actual_num >= expected_num;
        if (op == "lt") return actual_num < expected_num;
        if (op == "lte") return actual_num <= expected_num;
    }

    if (op == "regex" || op == "matches") {
        if (expected.IsString()) {
            try {
                std::regex re(expected.GetString());
                return std::regex_search(actual_value, re);
            } catch (const std::regex_error&) {
                spdlog::warn("Invalid regex in flag rule: {}", expected.GetString());
                return false;
            }
        }
        return false;
    }

    spdlog::debug("Unknown operator '{}' in flag rule", op);
    return false;
}

bool FlagEvaluator::matchesRules(const std::string& rules_json, const EvalContext& ctx) {
    if (rules_json.empty() || rules_json == "[]") {
        return true; // No rules means the flag applies to everyone
    }

    rapidjson::Document doc;
    doc.Parse(rules_json.c_str(), rules_json.size());

    if (doc.HasParseError() || !doc.IsArray()) {
        spdlog::warn("Failed to parse rules JSON");
        return true; // Fail open -- if rules can't be parsed, apply the flag
    }

    const auto& rules = doc.GetArray();
    if (rules.Size() == 0) return true;

    // Rules are OR'd at the top level. Each rule can have conditions that
    // are AND'd together.
    for (rapidjson::SizeType i = 0; i < rules.Size(); ++i) {
        const auto& rule = rules[i];
        if (!rule.IsObject()) continue;

        // A rule has "conditions" (array of conditions, AND'd together)
        if (rule.HasMember("conditions") && rule["conditions"].IsArray()) {
            const auto& conditions = rule["conditions"].GetArray();
            bool all_match = true;

            for (rapidjson::SizeType j = 0; j < conditions.Size(); ++j) {
                if (!matchesCondition(conditions[j], ctx)) {
                    all_match = false;
                    break;
                }
            }

            if (all_match) return true;
        }
        // Also support a flat rule that is itself a single condition
        else if (rule.HasMember("attribute") && rule.HasMember("operator")) {
            if (matchesCondition(rule, ctx)) return true;
        }
    }

    return false; // No rules matched
}

EvalResult FlagEvaluator::evaluate(const FlagConfig& flag, const EvalContext& ctx) {
    EvalResult result;
    result.key = flag.key;

    // If flag is disabled, return the default value
    if (!flag.enabled) {
        result.enabled = false;
        result.value = flag.default_value;
        result.reason = "disabled";
        return result;
    }

    // Determine user identifier for hashing
    std::string user_id = ctx.user_id.empty() ? ctx.anonymous_id : ctx.user_id;
    if (user_id.empty()) {
        result.enabled = false;
        result.value = flag.default_value;
        result.reason = "error";
        return result;
    }

    // Check targeting rules
    if (!matchesRules(flag.rules_json, ctx)) {
        result.enabled = false;
        result.value = flag.default_value;
        result.reason = "rule_no_match";
        return result;
    }

    // Check rollout percentage
    if (!isInRollout(flag.key, user_id, flag.rollout_percentage)) {
        result.enabled = false;
        result.value = flag.default_value;
        result.reason = "rollout";
        return result;
    }

    // Flag is active for this user
    result.enabled = true;
    result.reason = "rule_match";

    // Determine the value based on variant_type
    if (flag.variant_type == "boolean") {
        result.value = "true";
    } else if (!flag.variants_json.empty() && flag.variants_json != "[]") {
        // Multi-variate flag -- select variant
        std::string variant = selectVariant(flag.key, user_id, flag.variants_json);
        if (!variant.empty()) {
            result.value = variant;
            result.variant = variant;
        } else {
            result.value = flag.default_value;
        }
    } else {
        result.value = flag.default_value;
        result.reason = "default";
    }

    return result;
}

std::vector<EvalResult> FlagEvaluator::evaluateAll(const std::vector<FlagConfig>& flags,
                                                     const EvalContext& ctx) {
    std::vector<EvalResult> results;
    results.reserve(flags.size());

    for (const auto& flag : flags) {
        results.push_back(evaluate(flag, ctx));
    }

    return results;
}

} // namespace apdl

File: services/config/app/flags/evaluator.py
"""Flag evaluation engine.

Implements FNV-1a hashing for consistent bucketing, percentage-based rollouts,
multi-variate selection, and rule-based targeting. Byte-identical to the C++
FlagEvaluator for hash bucketing.
"""

import json
import logging
import re

logger = logging.getLogger(__name__)

_UINT32_MAX = 0xFFFFFFFF


def hash_bucket(key: str, user_id: str) -> int:
    """FNV-1a 32-bit hash for consistent bucketing.

    Deterministic and uniform distribution, suitable for percentage-based
    rollouts and experiment variant assignment.
    """
    data = f"{key}:{user_id}"
    h = 2166136261
    for c in data.encode("utf-8"):
        h ^= c
        h = (h * 16777619) & _UINT32_MAX
    return h


def is_in_rollout(flag_key: str, user_id: str, percentage: float) -> bool:
    """Check if a user falls within the rollout percentage."""
    if percentage >= 100.0:
        return True
    if percentage <= 0.0:
        return False
    h = hash_bucket(flag_key, user_id)
    bucket = (h / _UINT32_MAX) * 100.0
    return bucket < percentage


def select_variant(flag_key: str, user_id: str, variants_json: str) -> str:
    """Select a variant based on user hash and variant weights.

    Matches C++ selectVariant exactly.
    """
    try:
        variants = json.loads(variants_json)
    except (json.JSONDecodeError, TypeError):
        return ""

    if not isinstance(variants, list) or len(variants) == 0:
        return ""

    # Calculate total weight
    total_weight = 0.0
    for v in variants:
        if isinstance(v, dict) and "weight" in v:
            total_weight += float(v["weight"])

    if total_weight <= 0.0:
        # Equal weights if none specified
        total_weight = float(len(variants))

    h = hash_bucket(flag_key + ":variant", user_id)
    bucket = (h / _UINT32_MAX) * total_weight

    cumulative = 0.0
    for i, v in enumerate(variants):
        if not isinstance(v, dict):
            continue

        weight = 1.0  # default equal weight
        if "weight" in v and isinstance(v["weight"], (int, float)):
            weight = float(v["weight"])

        cumulative += weight
        if bucket < cumulative:
            # Return this variant's value
            if "value" in v:
                val = v["value"]
                if isinstance(val, str):
                    return val
                # Serialize non-string values (match C++ behavior)
                return json.dumps(val, separators=(",", ":"))
            if "key" in v and isinstance(v["key"], str):
                return v["key"]
            return str(i)

    # Fallback (should not happen with proper weights)
    return ""


def matches_condition(condition: dict, ctx: dict) -> bool:
    """Check a single rule condition against the evaluation context.

    Supports all operators from the C++ implementation including aliases.
    """
    if not isinstance(condition, dict):
        return False

    attribute = condition.get("attribute")
    op = condition.get("operator")
    if not isinstance(attribute, str) or not isinstance(op, str):
        return False

    # Look up the attribute value from the context
    if attribute in ("user_id", "userId"):
        actual_value = ctx.get("user_id", "")
    elif attribute in ("anonymous_id", "anonymousId"):
        actual_value = ctx.get("anonymous_id", "")
    else:
        attributes = ctx.get("attributes", {})
        if attribute in attributes:
            actual_value = attributes[attribute]
        else:
            # Attribute not found -- most operators should not match
            if op in ("not_exists", "is_not_set"):
                return True
            return False

    # Operators that don't need a value
    if "value" not in condition:
        if op in ("exists", "is_set"):
            return bool(actual_value)
        if op in ("not_exists", "is_not_set"):
            return not actual_value
        return False

    expected = condition["value"]

    # equals / eq / is
    if op in ("equals", "eq", "is"):
        if isinstance(expected, str):
            return actual_value == expected
        if isinstance(expected, bool):
            return actual_value == ("true" if expected else "false")
        if isinstance(expected, int):
            return actual_value == str(expected)
        if isinstance(expected, float):
            return actual_value == str(expected)
        return False

    # not_equals / neq / is_not
    if op in ("not_equals", "neq", "is_not"):
        if isinstance(expected, str):
            return actual_value != expected
        if isinstance(expected, bool):
            return actual_value != ("true" if expected else "false")
        return True

    # contains
    if op == "contains":
        if isinstance(expected, str):
            return expected in actual_value
        return False

    # not_contains
    if op == "not_contains":
        if isinstance(expected, str):
            return expected not in actual_value
        return True

    # starts_with
    if op == "starts_with":
        if isinstance(expected, str):
            return actual_value.startswith(expected)
        return False

    # ends_with
    if op == "ends_with":
        if isinstance(expected, str):
            return actual_value.endswith(expected)
        return False

    # in
    if op == "in":
        if isinstance(expected, list):
            for item in expected:
                if isinstance(item, str) and actual_value == item:
                    return True
        return False

    # not_in
    if op == "not_in":
        if isinstance(expected, list):
            for item in expected:
                if isinstance(item, str) and actual_value == item:
                    return False
        return True

    # Numeric comparisons: gt, gte, lt, lte
    if op in ("gt", "gte", "lt", "lte"):
        try:
            actual_num = float(actual_value)
        except (ValueError, TypeError):
            return False

        if isinstance(expected, (int, float)):
            expected_num = float(expected)
        elif isinstance(expected, str):
            try:
                expected_num = float(expected)
            except (ValueError, TypeError):
                return False
        else:
            return False

        if op == "gt":
            return actual_num > expected_num
        if op == "gte":
            return actual_num >= expected_num
        if op == "lt":
            return actual_num < expected_num
        if op == "lte":
            return actual_num <= expected_num

    # regex / matches
    if op in ("regex", "matches"):
        if isinstance(expected, str):
            try:
                return bool(re.search(expected, actual_value))
            except re.error:
                logger.warning("Invalid regex in flag rule: %s", expected)
                return False
        return False

    logger.debug("Unknown operator '%s' in flag rule", op)
    return False


def matches_rules(rules_json: str, ctx: dict) -> bool:
    """Check if the context matches any of the targeting rules.

    Rules are OR'd at the top level. Conditions within a rule are AND'd.
    Empty rules match everyone.
    """
    if not rules_json or rules_json == "[]":
        return True

    try:
        rules = json.loads(rules_json)
    except (json.JSONDecodeError, TypeError):
        logger.warning("Failed to parse rules JSON")
        return True  # Fail open

    if not isinstance(rules, list) or len(rules) == 0:
        return True

    for rule in rules:
        if not isinstance(rule, dict):
            continue

        # A rule has "conditions" (array of conditions, AND'd together)
        if "conditions" in rule and isinstance(rule["conditions"], list):
            conditions = rule["conditions"]
            all_match = True
            for cond in conditions:
                if not matches_condition(cond, ctx):
                    all_match = False
                    break
            if all_match:
                return True
        # Also support a flat rule that is itself a single condition
        elif "attribute" in rule and "operator" in rule:
            if matches_condition(rule, ctx):
                return True

    return False


def evaluate(flag: dict, ctx: dict) -> dict:
    """Evaluate a single flag against the given context.

    Returns a dict with keys: key, enabled, value, variant, reason.
    Reasons: "disabled", "error", "rule_no_match", "rollout", "rule_match", "default".
    """
    result = {
        "key": flag.get("key", ""),
        "enabled": False,
        "value": "",
        "variant": "",
        "reason": "",
    }

    # If flag is disabled, return the default value
    if not flag.get("enabled", False):
        result["value"] = flag.get("default_value", "false")
        result["reason"] = "disabled"
        return result

    # Determine user identifier for hashing
    user_id = ctx.get("user_id", "") or ctx.get("anonymous_id", "")
    if not user_id:
        result["value"] = flag.get("default_value", "false")
        result["reason"] = "error"
        return result

    # Check targeting rules
    if not matches_rules(flag.get("rules_json", "[]"), ctx):
        result["value"] = flag.get("default_value", "false")
        result["reason"] = "rule_no_match"
        return result

    # Check rollout percentage
    if not is_in_rollout(
        flag.get("key", ""), user_id, flag.get("rollout_percentage", 100.0)
    ):
        result["value"] = flag.get("default_value", "false")
        result["reason"] = "rollout"
        return result

    # Flag is active for this user
    result["enabled"] = True
    result["reason"] = "rule_match"

    variant_type = flag.get("variant_type", "boolean")
    variants_json = flag.get("variants_json", "[]")

    if variant_type == "boolean":
        result["value"] = "true"
    elif variants_json and variants_json != "[]":
        # Multi-variate flag -- select variant
        variant = select_variant(flag.get("key", ""), user_id, variants_json)
        if variant:
            result["value"] = variant
            result["variant"] = variant
        else:
            result["value"] = flag.get("default_value", "false")
    else:
        result["value"] = flag.get("default_value", "false")
        result["reason"] = "default"

    return result


def evaluate_all(flags: list[dict], ctx: dict) -> list[dict]:
    """Evaluate all flags against the given context."""
    return [evaluate(f, ctx) for f in flags]
